src/functional/ffront/foast_passes/closure_var_folding.py

```diff
@@ -0,0 +1,79 @@
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from typing import Any
+
+import functional.ffront.field_operator_ast as foast
+from eve import NodeTranslator, traits
+from eve.utils import FrozenNamespace
+
+
+class ClosureVarFolding(NodeTranslator, traits.VisitorWithSymbolTableTrait):
+    """
+    Replace `Name` nodes that refer to closure variable by `Constant` nodes.
+
+    `Attribute` nodes are also replaced by a `Constant` nodes with the value of the
+     closure variable's corresponding attribute. This is executed recursively for
+     attributes of attributes.
+    """
+
+    closure_vars: dict[str, Any]
+    _current_closure_vars: list[foast.Symbol]
+
+    def __init__(self, closure_vars: dict[str, Any]):
+        self.closure_vars = closure_vars
+
+    @classmethod
+    def apply(cls, node: foast.FieldOperator, closure_vars: dict[str, Any]):
+        return cls(closure_vars=closure_vars).visit(node)
+
+    def visit_Name(self, node: foast.Name, **kwargs):
+        symtable = kwargs["symtable"]
+        if node.id in symtable:
+            definition = symtable[node.id]
+            if definition in self._current_closure_vars:
+                value = self.closure_vars[node.id]
+                if isinstance(value, FrozenNamespace):
+                    return foast.Constant(value=value, location=node.location)
+        return node
+
+    def visit_Attribute(self, node: foast.Attribute, **kwargs):
+        # TODO: fix import form parent module by restructuring exception classis
+        from functional.ffront.func_to_foast import FieldOperatorSyntaxError
+
+        value = self.visit(node.value, **kwargs)
+        if isinstance(value, foast.Constant):
+            if hasattr(value.value, node.attr):
+                return foast.Constant(value=getattr(value.value, node.attr), location=node.location)
+            # TODO: fix this by restructuring exception classis
+            raise FieldOperatorSyntaxError(
+                msg="Constant does not have the attribute specified by the AST.",
+                filename=node.location.source,
+                lineno=node.location.line,
+                offset=node.location.column,
+                end_lineno=node.location.end_line,
+                end_offset=node.location.end_column,
+            )
+        # TODO: fix this by restructuring exception classis
+        raise FieldOperatorSyntaxError(
+            msg="Attribute can only be used on constants.",
+            filename=node.location.source,
+            lineno=node.location.line,
+            offset=node.location.column,
+            end_lineno=node.location.end_line,
+            end_offset=node.location.end_column,
+        )
+
+    def visit_FunctionDefinition(self, node: foast.FunctionDefinition, **kwargs):
+        self._current_closure_vars = node.closure_vars
+        return self.generic_visit(node, **kwargs)
```

Comment:
```suggestion
        return self.generic_visit(node, closure_vars=node.closure_vars ,**kwargs)
```
Avoids making pass stateful.

Reply:
Done